Test for broken JAR (zip) files

lkiesow opened the pull request:
This ensures that our JAR files are actually valid zip files since we seem to have broken files headers in several releases and this is causing issues with recent JVM updates.

This is not a very nice fix, but it should do the trick. It seems like Opencast 15 no longer suffers from this bug. We can probably remove this workaround in that branch. Though I would maybe keep the integrity test around for a bit longer.

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

lkiesow committed:
Fix broken JAR (zip) files

This ensures that our JAR files are actually valid zip files since we
seem to have broken files headers in several releases and this is
causing issues with recent JVM updates.

This is not a very nice fix, but it should do the trick. It seems like
Opencast 15 no longer suffers from this bug. We can probably remove this
workaround in that branch. Though I would maybe keep the integrity test
around for a bit longer.

ziegenberg commented on .github/fix-jars:
```diff
@@ -0,0 +1,33 @@
+
+  find . -name '*jar' | while read i
+  do
+    if ! zip --test "$i" &> /dev/null
```
ShellCheck says: In POSIX sh, `&>` is undefined. Use the long form instead of the shorthand:
```suggestion
    if ! zip --test "$i" > /dev/null 2>&1
```

`&>` is a bash and ksh extension for redirecting both stdout and stderr. In dash and POSIX sh, use `2>&1` explicitly.

ziegenberg commented on .github/fix-jars:
```diff
@@ -0,0 +1,33 @@
+      jarpath="$(readlink -f "$i")"
+      cd "${tmpdir}"
+      unzip "${jarpath}"
+      zip -r --compression-method store fix.zip *
```
Using `./*glob*` looks a bit safer.

```suggestion
      zip -r --compression-method store fix.zip ./*
```

ziegenberg commented on .github/fix-jars:
```diff
@@ -0,0 +1,33 @@
+
+  cd "${distname}/system/org/opencastproject/"
+
+  find . -name '*jar' | while read i
```
ShellCheck says: `read` without `-r` will mangle backslashes.


```suggestion
  find . -name '*jar' | while read -r i
```

ziegenberg commented on .github/fix-jars:
`fix-jars` does what says it should do. Run it locally to verify.

ziegenberg commented on .github/test-jars:
```diff
@@ -0,0 +1,8 @@
+
+set -eu
+
+find build/opencast-*/system/org/opencastproject -name '*.jar' | while read jarfile
```
ShellCheck says: `read` without `-r` will mangle backslashes.
```suggestion
find build/opencast-*/system/org/opencastproject -name '*.jar' | while read -r jarfile
```

ziegenberg commented on .github/test-jars:
`test-jars` does what says it should do. Run it locally to verify.

ziegenberg reviewed (changes requested):
I guess it looks fine, just a minor suggestion. Would be nice to have others' approval too.

lkiesow commented:
Thanks for the review. Applied all changes.

ziegenberg commented:
Maybe PR #5158 makes this obsolete?

ziegenberg commented:
@lkiesow As most certainly #5158 will find it's into the codebase I suggest throwing `github/fix-jars` overboard and just keeping the verify integrity bits. That shouldn't hurt us and might prevent this from happening again in the future.

lkiesow committed:
Don't repack JARs

Since we fixed the problem of having broken JAR files by updating the
maven-bundle-plugin, we don't need to repack the JAR files any longer.

But to make sure the problem does not persist, this patch still keeps
the test for checking that the JARs are actually good.

lkiesow commented:
> @lkiesow As most certainly #5158 will find it's into the codebase I suggest throwing `github/fix-jars` overboard and just keeping the verify integrity bits. That shouldn't hurt us and might prevent this from happening again in the future.

Done. This pull request now just contains the test. Nothing else. Can you do the review/merge?

ziegenberg commented:
Looks good to me

JamesUoM commented:
As java use to happily parse these jar files in the previous point release it is considered a bug that has been fixed in 11 and 17 https://bugs.openjdk.org/browse/JDK-8313765
